### PR TITLE
TRON-36 - trx recoveries, initial

### PR DESCRIPTION
**Pull request by twtaylorbitgo**
Initial PR for tron wallet recoveries. Still a WIP.

### CODE REVIEW

File: modules/core/src/v2/coins/trx.ts [hunk 1]
```diff
@@ -29,6 +31,9 @@ import { BitGo } from '../../bitgo';
 import { NodeCallback } from '../types';
 import { TransactionBuilder } from '@bitgo/account-lib';
 
+import debug = require('debug');
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Import is not used, please delete

File: modules/core/src/v2/coins/trx.ts [hunk 2]
```diff
@@ -29,6 +31,9 @@ import { BitGo } from '../../bitgo';
...
+import {utxo} from "@bitgo/statics/dist/src/utxo";
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Import is not used, please delete

File: modules/core/src/v2/coins/trx.ts [hunk 3]
```diff
@@ -29,6 +31,9 @@ import { BitGo } from '../../bitgo';
 import { NodeCallback } from '../types';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Duplicate import of bitgo-utxo-lib on line 29

File: modules/core/src/v2/coins/trx.ts [hunk 4]
```diff
@@ -29,6 +31,9 @@ import { BitGo } from '../../bitgo';
...
 import { TransactionBuilder } from '@bitgo/account-lib';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Duplicate import of @bitgo/account-lib

File: modules/core/src/v2/coins/trx.ts [hunk 5]
```diff
@@ -7,6 +7,8 @@ import { CoinFamily } from '@bitgo/statics';
 const co = Bluebird.coroutine;
 import * as bitgoAccountLib from '@bitgo/account-lib';
 import { HDNode } from 'bitgo-utxo-lib';
+import * as request from 'superagent';
+import * as common from '../../common';
 
 import { BaseCoin as StaticsBaseCoin } from '@bitgo/statics';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Duplicate import of @bitgo/statics

File: modules/core/src/v2/coins/trx.ts [hunk 6]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+  xprvToCompressedPrv(xprv: string): string {
+    if (!this.isValidXprv(xprv)) {
+      throw new Error('invalid xprv');
+    }
+
+    const hdNode = utxoLib.HDNode.fromBase58(xprv, this.bitcoinEncoding());
+    return hdNode.keyPair.d.toBuffer(32).toString('hex');
+  };
+
+  bitcoinEncoding(): any {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Remove `any`, return type can be inferred.

Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Make private.

File: modules/core/src/v2/coins/trx.ts [hunk 7]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+  /**
+   * Converts an xpub to a compressed pub
+   * @param xpub
+   */
+  xpubToCompressedPub(xpub: string): string {
+    if (!this.isValidXpub(xpub)) {
+      throw new Error('invalid xpub');
+    }
+
+    const hdNode = utxoLib.HDNode.fromBase58(xpub, this.bitcoinEncoding());
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: What's going on here? Why are we providing our own network object?

File: modules/core/src/v2/coins/trx.ts [hunk 8]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+    };
+  }
+
+  /**
+   * Make a query to Trongrid for information such as balance, token balance, solidity calls
+   * @param query {Object} key-value pairs of parameters to append after /api
+   * @param callback
+   * @returns {Object} response from Trongrid
+   */
+  recoveryPost(query: { path: string, jsonObj: any, node: NodeTypes }, callback?: NodeCallback<any>): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Make private

Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Can a more specific return type be used here instead of any?

File: modules/core/src/v2/coins/trx.ts [hunk 9]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+        .asCallback(callback);
+  }
+
+  /**
+   * Query our explorer for the balance of an address
+   * @param address {String} the address encoded in hex
+   * @param callback
+   * @returns {BigNumber} address balance
+   */
+  getAccountBalanceFromNode(address: string, callback?: NodeCallback<any>): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Can a more specific return type be used here instead of any?

File: modules/core/src/v2/coins/trx.ts [hunk 10]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+  }
+
+  /**
+   * Retrieves our build transaction from a node.
+   * @param toAddr hex-encoded address
+   * @param fromAddr hex-encoded address
+   * @param amount
+   * @param callback
+   */
+  getBuildTransaction(toAddr: string, fromAddr: string, amount: number, callback?: NodeCallback<any>): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Make private if not meant for external usage. Also can we use a more specific return type?

File: modules/core/test/v2/unit/recovery.ts [hunk 11]
```diff
@@ -554,6 +555,36 @@ describe('Recovery:', function() {
...
+      const recoveryTx = yield baseCoin.recover({
+        userKey: '{"iv":"QPX3xtGROshqHW8kGPAYCw==","v":1,"iter":10000,"ks":256,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"LJ4YYCyClRE=","ct":"hlJH8lWk/FaciymG8UsscxVFCnOduLRjoWxaK8xU7TjsqUDXsQjj0BpH7aNm64p6ldueaGoU2/VfrrzX9lWrcVmXspFp2oON5EyK45JbI13hirqG2dkOqoT8G8mrMydMp6zG5iOA+EtXRy69kYDCI1Re6mR7k1c="}',
+        backupKey: '{"iv":"xbOCFaZVnrQLAYKcgMvdNw==","v":1,"iter":10000,"ks":256,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"86cZ+hT+S0Y=","ct":"RH7Uks/JjNARX9wuIw6r4Map2C9FtLlWYk4zjLcrjzkBuNTCUNQgxF7kU5/SWzD+tomVBj6P9CLkXYzPlR1NhVi+aT9mTW6LK9nJ+ErpLKXzbIAxBLezDjJ5xqUS5cGkjoHCtANL7qTZcDBvOfejLDrUjQdw2WQ="}',
+        bitgoKey: 'xpub661MyMwAqRbcFcaxCPsEyhj79VUuVVThWinZnjhvAPnFLB1SBp7Yk4gvqWsGE3MHdw1tPRLnHRRQLNcrKqaCyBnFK5XTrZUrLyY94LXn4v9',
+        walletPassphrase: TestBitGo.V2.TEST_RECOVERY_PASSCODE,
+        recoveryDestination: 'TYBTURKpanKxnx91uyfvvtztNeHE3EQf6G',
+      });
+    }));
+
+    it('should generate TRON recovery tx with unencrypted keys', co(function *() {
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/recovery.ts]: Please complete test or remove

File: modules/core/test/v2/unit/recovery.ts [hunk 12]
```diff
@@ -554,6 +555,36 @@ describe('Recovery:', function() {
...
+
+    }));
+
+    it('should generate an TRON unsigned sweep', co(function *() {
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/recovery.ts]: Please complete test or remove

File: modules/core/src/v2/coins/trx.ts [hunk 13]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+  }
+
+  xprvToCompressedPrv(xprv: string): string {
+    if (!this.isValidXprv(xprv)) {
+      throw new Error('invalid xprv');
+    }
+
+    const hdNode = utxoLib.HDNode.fromBase58(xprv, this.bitcoinEncoding());
+    return hdNode.keyPair.d.toBuffer(32).toString('hex');
+  };
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Unnecessary semicolon

File: modules/core/src/v2/coins/trx.ts [hunk 14]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+      // get our user, backup keys
+      const keys = yield self.initiateRecovery(params);
+
+      // we need to decode our bitgoKey to a base58 address
+      const bitgoAddress = self.compressedPubToHexAddress(self.xpubToCompressedPub(params.bitgoKey));
+      const recoveryAddressHex = bitgoAccountLib.Trx.Utils.getHexAddressFromBase58Address(params.recoveryDestination);
+
+      const recoveryAmount = yield self.getAccountBalanceFromNode(bitgoAddress);
+
+      //const userPrv = HDKey.fromExtendedKey(keys[0]).privateKey;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Remove dead code

File: modules/core/src/v2/coins/trx.ts [hunk 15]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+      const userXPub = keys[0].neutered().toBase58();
+      const userXPrv = keys[0].toBase58();
+      const backupXPub = keys[1].neutered().toBase58();
+
+      const userPrv = self.xprvToCompressedPrv(userXPrv);
+      const userHexAddr = self.compressedPubToHexAddress(self.xpubToCompressedPub(userXPub));
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Variable is not used, please delete

File: modules/core/src/v2/coins/trx.ts [hunk 16]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+      const backupHexAddr = self.compressedPubToHexAddress(self.xpubToCompressedPub(backupXPub));
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Variable is not used, please delete

File: modules/core/src/v2/coins/trx.ts [hunk 17]
```diff
@@ -249,6 +275,200 @@ export class Trx extends BaseCoin {
...
+
+      // construct the tx
+      // TODO: recovery amount might not include fees, we'll have to figure out the max we can spend
+      //   sensitive to energy and tx cost amounts
+      // there's an assumption here being made about fees: for a wallet that hasn't been used in awhile, the implication is
+      // it has maximum bandwidth. thus, a recovery should cost the minimum amount (1e6 sun)
+      if (1e6 > recoveryAmount) {
+        throw new Error('Amount of funds to recover wouldnt be able to fund a send');
+      }
+      const recoveryAmountMinusFees = recoveryAmount - 1e6;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/trx.ts]: Are we sure this is safe to hard code? Might this be changed in the future?

File: modules/core/src/v2/coins/trx.ts (hunk 4 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/trx.ts]: this was not me but I will remove it ;)

File: modules/core/src/v2/coins/trx.ts (hunk 6 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/trx.ts]: removed totally

File: modules/core/src/v2/coins/trx.ts (hunk 8 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/trx.ts]: it varies based on the route unfortunately

File: modules/core/src/v2/coins/trx.ts (hunk 17 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/trx.ts]: could be. I added a const for it. in theory should be sourced from account lib @argjv

**Pull request merged**